nav2_rviz_plugins/src/nav2_panel.cpp

```diff
@@ -93,6 +173,108 @@ Nav2Panel::onShutdown()
   QFuture<void> future =
     QtConcurrent::run(std::bind(&nav2_lifecycle_manager::LifecycleManagerClient::shutdown,
       &client_));
+  timer_->cancel();
+}
+
+void
+Nav2Panel::onCancel()
+{
+  QFuture<void> future =
+    QtConcurrent::run(std::bind(&Nav2Panel::onCancelButtonPressed,
+      this));
+}
+
+void
+Nav2Panel::onNewGoal(double x, double y, double theta, QString frame)
+{
+  auto pose = geometry_msgs::msg::PoseStamped();
+
+  pose.header.stamp = rclcpp::Clock().now();
+  pose.header.frame_id = frame.toStdString();
+  pose.pose.position.x = x;
+  pose.pose.position.y = y;
+  pose.pose.position.z = 0.0;
+  pose.pose.orientation = orientationAroundZAxis(theta);
+
+  startNavigation(pose);
+}
+
+
+geometry_msgs::msg::Quaternion
+Nav2Panel::orientationAroundZAxis(double angle)
+{
+  tf2::Quaternion q;
+  q.setRPY(0, 0, angle);  // void returning function
+  return tf2::toMsg(q);
+}
+
+void
+Nav2Panel::onCancelButtonPressed()
+{
+  auto future_cancel = action_client_->async_cancel_goal(goal_handle_);
+
+  if (rclcpp::spin_until_future_complete(client_node_, future_cancel) !=
+    rclcpp::executor::FutureReturnCode::SUCCESS)
+  {
+    RCLCPP_ERROR(client_node_->get_logger(), "Failed to cancel goal");
+    return;
+  }
+}
+
+void
+Nav2Panel::timerActionEvent()
+{
+  if (!goal_handle_) {
+    RCLCPP_DEBUG(client_node_->get_logger(), "Waiting for Goal");
+    state_machine_.postEvent(new ROSActionQEvent(false));
+    return;
+  }
+
+  rclcpp::spin_some(client_node_);
+  auto status = goal_handle_->get_status();
+
+  // Check if the goal is still executing
+  if (status == action_msgs::msg::GoalStatus::STATUS_ACCEPTED ||
+    status == action_msgs::msg::GoalStatus::STATUS_EXECUTING)
+  {
+    state_machine_.postEvent(new ROSActionQEvent(true));
+  } else {
+    state_machine_.postEvent(new ROSActionQEvent(false));
+  }
+}
+
+void
+Nav2Panel::startNavigation(geometry_msgs::msg::PoseStamped pose)
+{
+  auto is_action_server_ready = action_client_->wait_for_action_server(std::chrono::seconds(5));
+  if (!is_action_server_ready) {
+    RCLCPP_ERROR(client_node_->get_logger(), "NavigateToPose action server is not available."
+      " Is the initial pose set?");
+    return;
+  }
+
+  // Send the goal pose
+  goal_.pose = pose;
+
+  // Enable result awareness by providing an empty lambda function
```

Comment:
What's "result awareness"? Can you elaborate on this comment

Reply:
I'm not entirely sure as this wasn't my code/comment, just part of refactor, perhaps @mjeronimo could explain better.

Reply:
"Result awareness" was introduced into the ROS2 action client. There were some relatively recent changes (a couple months ago or so), where one now provides callbacks in the SendGoalOptions for the goal response, feedback, and result. The SendGoalOptions are then passed to the client's async_send_goal method. In the navigation stack we already had code written to the old style of explicitly sending the goal request, etc. This SendGoalOptions change broke our code. One way I found to work around it was to provide an empty callback for the result_callback. Providing this callback causes the ROS2 to set the result aware flag (whatever that means) and allows us to call the existing functions. Alternatively, we could update our client code to conform to the new callback model. See the ROS2 code in rclcpp_action for more about "result awareness." I haven't found any docs.